swift/Workflow/Sources/SubtreeManager.swift

```diff
@@ -9,29 +9,28 @@ extension WorkflowNode {
 
         internal var onUpdate: ((Output) -> Void)? = nil
 
-        /// Signals from the outside world (e.g. UI)
-        private let (childEvent, childEventObserver) = Signal<Signal<Output, NoError>, NoError>.pipe()
+        /// Sinks from the outside world (i.e. UI)
+        private var sinks: [SinkContext] = []
 
         /// The current array of children
         private (set) internal var childWorkflows: [ChildKey:AnyChildWorkflow] = [:]
 
+        /// The current array of workers
         private (set) internal var childWorkers: [AnyChildWorker] = []
 
-        private let (lifetime, token) = Lifetime.make()
+        /// Subscriptions from the outside world.
+        private var subscriptionContext: SubscriptionContext = SubscriptionContext(eventSources: [])
 
-        init() {
-            childEvent
-                .flatMap(.latest, { $0 })
-                .take(during: lifetime)
-                .observe(on: QueueScheduler.workflowExecution)
-                .observeValues { [weak self] event in
-                    self?.handle(output: event)
-                }
-        }
+        init() {}
 
         /// Performs an update pass using the given closure.
         func render<Rendering>(_ actions: (RenderContext<WorkflowType>) -> Rendering) -> Rendering {
 
+            /// Invalidate the previous sinks.
+            for sink in sinks {
+                sink.invalidate()
```

Comment:
What does invalidating a sink do?

Reply:
Just sets its internal state to "invalid". `sink` in this scope is the (poorly named currently) `SinkContext` that is actually wired into the subtree manager.

If it receives an event when it's in the `invalid` state, it will assert (fail fast) instead of silently dropping it.

Reply:
Ah, ok, so this is basically the equivalent of this line in a PR i've got open:

https://github.com/square/workflow/blob/798afacb46481a8998deff2c3b47784609fd0de4/kotlin/workflow-runtime/src/main/java/com/squareup/workflow/internal/WorkflowNode.kt#L125

Reply:
Yup! I assume that `cancel` in that context will make it assert if it's still used?

Reply:
Yep.